packages/axios/src/index.ts

```diff
@@ -1,52 +1,172 @@
 import http from 'node:http';
 import https from 'node:https';
 
-import type { IConfig } from '@cortec/config';
-import type { IContext, IModule } from '@cortec/types';
-import type { Axios, AxiosRequestConfig } from 'axios';
-import axios from 'axios';
+import { type IConfig, Config } from '@cortec/config';
+import type { INewrelic } from '@cortec/newrelic';
+import {
+  type IContext,
+  type IModule,
+  type ITrace,
+  Headers,
+} from '@cortec/types';
+import type * as A from 'axios';
+import axios, { isAxiosError } from 'axios';
+import capitalize from 'lodash.capitalize';
 
 export interface IAxiosConfig {
-  global?: AxiosRequestConfig;
+  global?: A.AxiosRequestConfig;
   api: {
-    [key: string]: AxiosRequestConfig;
+    [key: string]: A.AxiosRequestConfig;
   };
 }
 
-export interface IRequester {
-  api(name: string): Axios;
+interface TracedAxios extends A.Axios {
+  trace(ctx: ITrace): A.Axios;
 }
 
-export default class CortecAxios implements IModule, IRequester {
+export interface IAxios {
+  service(name: string, flags?: RequestFlags): TracedAxios;
+}
+
+export enum RequestFlags {
+  None = 0,
+  Notice4XX = 1 << 0,
+}
+
+function match(trait: RequestFlags, flag: RequestFlags): boolean {
+  return (trait & flag) === flag;
+}
+
+class ExternalServiceError extends Error {
+  constructor(service: string, err: A.AxiosError) {
+    super(
+      `service=${service} method=${err.request?.method} path=${
+        err.request?.path
+      } status=${err.response?.status} body=${JSON.stringify(
+        err.response?.data
+      )}`
+    );
+    this.name = 'ExternalServiceError';
+  }
+}
+
+export default class Axios implements IModule, IAxios {
   name = 'axios';
-  private instances: Map<string, Axios> = new Map();
+  private instances: Map<string, A.Axios> = new Map();
+  private services: string[];
+  private nr: INewrelic | undefined;
```

Comment:
Not sure if I'm just not up to speed on idiomatic TS this way, so maybe I'm off base, but explicitly allowing `undefined` as an inhabitant of a type that we control makes me uncomfortable. It has bad and weird semantics, could we check for it at the edges and then have this field be `nr?: INewrelic` instead, unless this is for serialization purposes?

Reply:
I don't think class properties can be defined with `nr?` therefore have to use `INewrelice|undefined`.